From b8cd2d9b6eaba42b7b7752e346b5f7ee61415a9e Mon Sep 17 00:00:00 2001
From: Baoquan He <bhe@redhat.com>
Date: Fri, 15 Aug 2014 07:27:22 -0400
Subject: [kernel] kexec: save PG_head_mask in VMCOREINFO

Message-id: <1408087644-31621-2-git-send-email-bhe@redhat.com>
Patchwork-id: 87782
O-Subject: [PATCH 1/3] kexec: save PG_head_mask in VMCOREINFO
Bugzilla: 1115240
RH-Acked-by: Vivek Goyal <vgoyal@redhat.com>
RH-Acked-by: Chao WANG <chaowang@redhat.com>

Resolves: bz1115240
https://bugzilla.redhat.com/show_bug.cgi?id=1115240

This is back ported from upstream.

commit b3acc56bfe1287c6b666e80edc70b89eea2a1a80
Author: Petr Tesarik <ptesarik@suse.cz>
Date:   Mon Jun 23 13:22:03 2014 -0700

    kexec: save PG_head_mask in VMCOREINFO

    To allow filtering of huge pages, makedumpfile must be able to identify
    them in the dump.  This can be done by checking the appropriate page
    flag, so communicate its value to makedumpfile through the VMCOREINFO
    interface.

    There's only one small catch.  Depending on how many page flags are
    available on a given architecture, this bit can be called PG_head or
    PG_compound.

    I sent a similar patch back in 2012, but Eric Biederman did not like
    using an #ifdef.  So, this time I'm adding a common symbol
    (PG_head_mask) instead.

    See https://lkml.org/lkml/2012/11/28/91 for the previous version.

    Signed-off-by: Petr Tesarik <ptesarik@suse.cz>
    Acked-by: Vivek Goyal <vgoyal@redhat.com>
    Cc: Eric Biederman <ebiederm@xmission.com>
    Cc: Paul Mackerras <paulus@samba.org>
    Cc: Fengguang Wu <fengguang.wu@intel.com>
    Cc: Benjamin Herrenschmidt <benh@kernel.crashing.org>
    Cc: Shaohua Li <shli@kernel.org>
    Cc: Alexey Kardashevskiy <aik@ozlabs.ru>
    Cc: Sasha Levin <sasha.levin@oracle.com>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Brew build:
https://brewweb.devel.redhat.com/taskinfo?taskID=7844251

Test: Passed on kvm guest.

Signed-off-by: Baoquan He <bhe@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/page-flags.h b/include/linux/page-flags.h
index 6d53675..5bc4f40 100644
--- a/include/linux/page-flags.h
+++ b/include/linux/page-flags.h
@@ -346,6 +346,9 @@ static inline void ClearPageCompound(struct page *page)
  ClearPageHead(page);
 }
 #endif
+
+#define PG_head_mask ((1L << PG_head))
+
 #else
 /*
  * Reduce page flag use as much as possible by overlapping
diff --git a/kernel/kexec.c b/kernel/kexec.c
index db1c571..650eef6 100644
--- a/kernel/kexec.c
+++ b/kernel/kexec.c
@@ -1680,6 +1680,7 @@ static int __init crash_save_vmcoreinfo_init(void)
 #ifdef CONFIG_MEMORY_FAILURE
  VMCOREINFO_NUMBER(PG_hwpoison);
 #endif
+ VMCOREINFO_NUMBER(PG_head_mask);
  VMCOREINFO_NUMBER(PAGE_BUDDY_MAPCOUNT_VALUE);
 
  arch_crash_save_vmcoreinfo();
-- 
1.7.1